lambdas/preview/index.py

```diff
@@ -141,35 +145,102 @@ def extract_vcf(file_):
     """
     Pull summary info from VCF: meta-information, header line, and data lines
     (in that order, up to MAX_LINES)
-    """
 
+    Args:
+        file_ - file-like
+    """
     meta = []
     header = []
     data = []
-
+    size = 0
     with open(file_.name, 'rb') as vcf:
         for index, line in enumerate(vcf, start=1):
-            # don't escape quotes
-            line = line.strip().decode('unicode_escape')
-            formatted = f'<p>{line}</p>'
+            line = _truncate(line.rstrip().decode(), size)
             if line.startswith('##'):
-                meta.append(formatted)
+                meta.append(line)
             elif line.startswith('#'):
-                header.append(formatted)
+                header.append(line)
             else:
-                data.append(formatted)
+                data.append(line)
+            size += len(line)
+            # stop if we're over the max
+            if index >= MAX_LINES or size > MAX_BYTES:
+                break
 
-            if index >= MAX_LINES:
+    info = {
+        'data': {
+            'meta': meta,
+            'header': header,
+            'data': data,
+        }
+    }
+
+    return '', info
+
+def extract_txt(file_):
+    """
+    Display first and last few lines of a potentially large file.
+
+    Args:
+        file_ - file-like object
+    """
+    size = 0
+    ellipsis = False
+    head = []
+    with open(file_.name, 'rb') as txt:
+        for index, line in enumerate(txt, start=1):
+            line = _truncate(line.rstrip().decode(), size)
+            size += len(line)
+            # this is a heuristic; can fail to return shorter tail lines
+            # if head lines are uncharacteristically long
+            # we're guarding against, for example, huge single-line JSON files
+            if index <= MAX_LINES and size <= MAX_BYTES:
+                head.append(line)
+            if size > MAX_BYTES:
                 break
+            if index > MAX_LINES:
+                ellipsis = True
```

Comment:
since this only happens in one case, you don't need a flag variable for ellipsis, you can just inline that whole method here and you don't need to reopen the file, immediately seek to the end and start your while loop. If you're worried about it being too much indentation you can use a named inner function instead (but cutting the `if ellipsis` and `with open` out means it shouldn't be too bad).

Reply:
yeah fixed - good feedback. all of the loops got tighter and shorter, so the local function declaration was not even needed